les/clientpool.go

```diff
@@ -398,6 +472,59 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
 			return f.connectedCap > f.capLimit || f.connectedQueue.Size() > f.connLimit
 		})
 	}
+	if f.eventHook != nil {
+		f.eventHook("capacityUpdate", nil)
+	}
+}
+
+// setCapacity sets the assigned capacity of a connected client
+func (f *clientPool) setCapacity(c *clientInfo, capacity uint64) error {
+	if f.connectedMap[c.id] != c {
+		return fmt.Errorf("client %064x is not connected", c.id[:])
+	}
+	if c.capacity == capacity {
+		return nil
+	}
+	if !c.priority {
+		return errNoPriority
+	}
+	oldCapacity := c.capacity
+	c.capacity = capacity
+	f.connectedCap += capacity - oldCapacity
+	c.balanceTracker.setCapacity(capacity)
+	f.connectedQueue.Update(c.queueIndex)
+	if f.connectedCap > f.capLimit {
+		var kickList []*clientInfo
+		kick := true
+		f.connectedQueue.MultiPop(func(data interface{}, priority int64) bool {
+			client := data.(*clientInfo)
+			kickList = append(kickList, client)
+			f.connectedCap -= client.capacity
+			if client == c {
+				kick = false
+			}
+			return kick && (f.connectedCap > f.capLimit)
+		})
+		if kick {
+			now := mclock.Now()
+			for _, c := range kickList {
+				f.dropClient(c, now, true)
+			}
+		} else {
+			c.capacity = oldCapacity
+			c.balanceTracker.setCapacity(oldCapacity)
```

Comment:
So we still need to set back `totalConnected`

Reply:
event subs are removed for now

Reply:
All clients in `kickList` are removed from `connectedCap`, including `c` if `kick` is false. They are added back after resetting `c.capacity = oldCapacity` so `connectedCap` should be correct in the end.